tutorials/4_arbitrary_clients.md

```diff
@@ -0,0 +1,261 @@
+in general, we must prove that adding an additional client will never introduce
+a new interaction which violates the property.
+
+We could tackle this problem directly by modeling every client. However, this
+would not scale to bundles with even a small amount of client state. Instead, it
+would be ideal to summarize the clients, and then retain only a small set of
+addresses. Even in the case of `Fund`, where clients lack state, we would still
+benefit from reducing the search space of interleaving client transactions.
+
+For this reason, let's try proving the property with a fixed number of
+addresses. For convenience, we will refer to the arrangement of clients and
+contracts as a *network topology*, and we will refer to the subset of clients as
+a *neighbourhood*. We will justify these terms in the next tutorial.
+
+So what happens when we construct a neighbourhood? If we select too few clients,
+we may lose behaviours from the original bundle (formally, this is an
+*underapproximation*). For instance, if we deployed `Fund` with a single client,
+the `owner` of `Fund` would never change. However, if we introduced too many
+clients, this could lead to redundant traces (formally, this is *symmetry*). For
+example, if we deployed `Fund` with 100 clients, there would be many identical
+ways to select an initial `owner`, and then transfer to a secondary `owner`.
+
+If we look closely at `Fund`, we will see that addresses only appear in equality
+relations. An equality relation can distinguish between any two address
+arguments, or between an address argument and a literal address. Note that due
+to the semantics of Solidity, `address(0)` is an implicit literal of all
+bundles. We are guaranteed to cover all paths of execution if we can assign a
+unique value to each address in any transaction. The specific values are
+unimportant, provided that we include all literals inside the neighbourhood.
+
+In other words, we must find the maximum number of addresses use by any
+transaction. We call this number the maximum *transactional address footprint*.
+As the `Manager` bundle is free from address operators, we only need to count
+the number of contract addresses, address variables, and address arguments in
+each transaction. We find that `Fund.releaseTo(address)` has the maximum address
+footprint, with two contract addresses, one address variable (`owner`) and two
+address arguments (`msg.sender` and `_new`). Therefore, our neighbourhood will
+have 6 addresses.
+
+### Instrumenting the Restricted Address Set
+
+SmartACE will construct the neighbourhood automatically. We can compare
+against SmartACE by runnings the following commands.
+
+  * `path/to/solc fund.sol --bundle=Manager --c-model --output-dir=fund`
+  * `cd fund ; mkdir build ; cd build`
+  * `cmake .. -DSEA_PATH=/path/to/seahorn/bin`
+  * `make run-clang-format`
+
+As before, we will open `../cmodel.c`. Let's start by inspecting the assignment
```

Comment:
use `cmake --build`

put the model online. create github repo with all files used by the tutorial and point there.